### PR TITLE
[BEAM-3308] Revamp Go SDK runtime

### PR DESCRIPTION
**Pull request by herohde**
(Please ignore the symbol caching commit. I'll rebase it away when #4256 is merged.)

### CODE REVIEW

**Comment by herohde** [comment]
R: @wcn3

File: sdks/go/pkg/beam/runners/direct/direct.go [hunk 1]
```diff
@@ -34,229 +34,233 @@ func Execute(ctx context.Context, p *beam.Pipeline) error {
...
+		for i := 0; i < len(edge.Input); i++ {
+			b.links[linkID{edge.ID(), i}] = u
+		}
+
+	case graph.DataSink:
+		u = &exec.DataSink{UID: b.idgen.New(), Edge: edge}
+
 	default:
-		panic(fmt.Sprintf("Unit %v has no output", unit))
+		return nil, fmt.Errorf("unexpected edge: %v", edge)
```
Comment (wcn3) [inline sdks/go/pkg/beam/runners/direct/direct.go]: Brainstorming about the beam.External issue, is it worth building an "unknown node" and linking it in here, and returning the unknown nodes as a return value of this function? That way, a caller could verify all the unknown nodes are beam.External nodes and be happy with the translation. As is, this code fails on the first translation error, and the caller won't be able to detect if there are other issues in the graph.

Comment (herohde) [reply sdks/go/pkg/beam/runners/direct/direct.go]: I think the External edges should be resolved/expanded before or at the time we reach this point. Although we're using the Compile function in the harness as well, I was rather thinking it should be a private part of the direct runner (see BEAM-3316).

**Comment by herohde** [comment]
Thanks!

**Comment by herohde** [comment]
R: @aaltay (merge)

**Pull request merged**